### ISSUE TITLE
Illumination patterns

### ISSUE THREAD

**Issue by lucarizzi** [Keck-DataReductionPipelines/KCWI_DRP#13]
From Ryan Cook:
 it seems there is a clear spatial pattern in KCWI:

![image](https://user-images.githubusercontent.com/524957/81980534-7fe70380-95ca-11ea-9016-4be9f821ff32.png)

**Comment by lucarizzi** [comment]
Almost seems like two patterns... a diagonal pattern across the entire frame and a horizontal/bent pattern towards the bottom...
The ones down the bottom... that's where the line of interest falls in our data
The features are likely physical, meaning part of the detector. 

In fact they are not visible in the pixel flat:

![image (1)](https://user-images.githubusercontent.com/524957/81980760-e704b800-95ca-11ea-9bfb-789d59076d94.png)

 the diagonal features (in the relative scaling image I posted earlier) are not currently being taken into account. (edited) 
Note, the pixelflat uses the same input images as the illumflat. So, as far as I know, there isn't a KCWI pixelflat with the detector uniformly illuminated...

**Comment by lucarizzi** [comment]
Contribution by joe hennawi:
to be more precise with language, we are not assuming a uniformly illuminated CCD, but rather we assume that the illumination profile does not depend on wavelength

I think I see your point about the coarse spectral profile, but I also fear that it may not work because the choice of spectral b-spline spacing may be arbitrary, and it is always better to perform a fit on your actual data

My suggestion is to figure out the variables in which this signal is coherent. For a single slit it is obviously (spec, spat) pixel position.

For many slits, you want to use wavelength instead of spec
8:47
So for KCWI, what is the second variable, is it spatial position?

SLit number?

If you can figure that out I can help you construct a 2-d model fit with the b-spline

that will work better than hacking things off of the flats because you are fitting the data to an actual model

**Comment by lucarizzi** [comment]
As a way of an update, here is the latest image of the wavy features in KCWI (BH2)
![image (2)](https://user-images.githubusercontent.com/524957/81980895-2a5f2680-95cb-11ea-9b3d-7e5761f3ee3c.png)
 think this shows diagonal features (high frequency), and some horizontal bands (low frequency). I've also aligned things in wavelength, and that doesn't seem to produce more coherent structures. So, I'm willing to bet that this is either scattered light or something physical with the pixels. A bit strange that it's so regular... (edited) 

There doesn't appear to be any diagonal features or horizontal bands in the arcs though (nor in the off slit regions of the science frame), so I'm resistant to dividing these features into the science frames. Ergo, the 2D bspline flat routine is essential for KCWI. (edited) 

I've produced some relative efficiency curves for the slits that don't include these features, and that seems to be working well for the sky-subtraction. Again, the 2D bspline skysub routine is phenomenal in PypeIt... one satisfied customer  :slightly_smiling_face:

**Comment by lucarizzi** [comment]
COntribution by Joe Hennawi:
Hi @Ryan Cooke if you don’t plan to divide this into the data what is the point of computing it?
I don’t see why this would be present only in the flats but not in the data unless it is a scattered light issue
Even then that does not make sense
scattered light is a function of the optics
the intensity levels just allow you to see it better
the only way this could “not be in the data” would be if the illumination pattern of the flat field lamps is different from the illumination pattern at infinity
I don’t know what your stretch is, so I cannot gauge how large these fluctuations are
Is this 1 percent or 10%?
I think if you stretch the data hard enough you can see these
I’ve definitely seen suck things in other instruments
If your goal is to construct a pixelflat, we should think about how to mix together different such patterns to average them down to be more smoth
In other words, can we shift the grating around or average flats from different setups?


I was calculating it to see what the key variables are that determine the shape. Note, this is also seen in the off slit parts of the detector. I'm guessing that it's scattered light, but I was surprised that I saw nothing in the arc lamp frames. (edited) 

Perhaps I could try the 1D sky sub on an Arc frame to see if there are any streaks there (but I don't see any streaks off the slits).

so, I'm not expecting it to show anything.

**Comment by lucarizzi** [comment]
Here is some for the BM grating:
![image (3)](https://user-images.githubusercontent.com/524957/81981057-73af7600-95cb-11ea-9443-f1697fdeaec5.png)
Similar diagonal pattern, but definitely different. So it's probably setup dependent.

Note, the peak-to-peak in this case (towards the bottom of the frame) is now on the order of 5-10%

**Comment by MateuszKM** [comment]
How long after a detector power cycle were these data taken?
